### PR TITLE
fix: correct Version and GitHash output for manager-api command

### LINKED ISSUE #1155: bug(cmd): Inaccurate output Version number of manger-api command
# Bug report

## Describe the bug

When running `manager-api` command generated by the latest code from `master` branch, the output `Version` is `2.2`, which is obviously inaccurate. The log shows as below:

```
➜  apisix-dashboard/output git:(master) ./manager-api
The manager-api is running successfully!

Version : 2.2
Listen  : 127.0.0.1:9000
Loglevel: warn
Logfile : /Users/imjoey/Work/apache-apisix/apisix-dashboard/output/logs/error.log
```

## How to Reproduce

1. Fetch the latest code from master branch;
2. Run `make build`;
3. Go into `output` directory and run `./manager-api`;
4. See error;

## Expected behavior

The `Version` should be `master`. 

## System information

- OS: macOS
- Version: master

## Additional context

While honestly, the `master` is still **not intuitive**, so shall we also add Git hash into the output? Such as:

```
Version : master
GitHash: 4317a5f
Listen  : 127.0.0.1:9000
Loglevel: warn
Logfile : /Users/imjoey/Work/apache-apisix/apisix-dashboard/output/logs/error.log
```

Any feedback is welcome. Thanks.

### PR DESCRIPTION
**Pull request by imjoey**
Signed-off-by: imjoey <majunjiev@gmail.com>

Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance

- Related issues

Fixes #1155 .

___
### Bugfix
- Description

Wrong version is put in `VERSION` file, and cannot support showing out the exact commit of code base.

- How to fix?

- Use `master` as the version number in `VERSION` file in `master` branch
- Add support for showing the git commit hash of the code base which is for generating the `manager-api` command

### CODE REVIEW

**Comment by imjoey** [comment]
@juzhiyuan @starsz I've no idea why the CI test *Semantic Pull Request* fails. Much appreciated if you can help me out. 😄  Thanks.

**Review by nic-chen — CHANGES_REQUESTED** [review]
Apache release can't include .git, so the build will fail. 

So if we want to add git version, we can add another script and add it there.

**Comment by juzhiyuan** [comment]
> @juzhiyuan @starsz I've no idea why the CI test _Semantic Pull Request_ fails. Much appreciated if you can help me out. 😄 Thanks.

image

just ignore this check, it's required to have at least 2 commits when using Squash 😂

File: api/build.sh [hunk 1]
```diff
@@ -20,6 +20,7 @@ export ENV=local
 pwd=`pwd`
 
 VERSION=$(cat ./api/VERSION)
+GITHASH=$(git log --pretty=format:"%h" -1)
```
Comment (juzhiyuan) [inline api/build.sh:23]: oh yes, if a user is using a release tar package from Apache Release, the package will not include `.git` folder.

cc @membphis any ideas?

Comment (membphis) [reply api/build.sh:23]: For the version released by apache, we add a new separate file to store the Git hash value.

Here is an example:

```
$ cat ./.githash
bae2e05b
```

Comment (membphis) [reply api/build.sh:23]: @nic-chen what do you think?

Comment (membphis) [reply api/build.sh:23]: and I think we generate the file automatically:

https://github.com/apache/apisix-dashboard/blob/master/Makefile#L96

Comment (tokers) [reply api/build.sh:23]: It seems ingress apisix also has this issue.

Comment (nic-chen) [reply api/build.sh:23]: > For the version released by apache, we add a new separate file to store the Git hash value.
> 
> Here is an example:
> 
> ```
> $ cat ./.githash
> bae2e05b
> ```

LGTM

Comment (imjoey) [reply api/build.sh:23]: @membphis @juzhiyuan @nic-chen @tokers

Thank you all for reviewing and suggestions. How about try to get the git hash from `./.githash` first, if that file does not exist, we use the `.git` folder?

Comment (imjoey) [reply api/build.sh]: @membphis @juzhiyuan @tokers @nic-chen 

Much appreciated for your suggestions. PR already is updated and adds support for the new `.githash` file. Ready for review. Thanks.

**Comment by nic-chen** [comment]
@imjoey 

It would be great to add an assertion for the modified code in the corresponding test case:
https://github.com/apache/apisix-dashboard/blob/master/api/test/shell/cli_test.sh#L118-L145

Thanks!

**Comment by imjoey** [comment]
> @imjoey
> 
> It would be great to add a test case for the modified code, please refer to:
> https://github.com/apache/apisix-dashboard/blob/master/api/test/shell/cli_test.sh#L118-L145
> 
> Thanks!

@nic-chen  sorry, my bad, thanks for your reminder. It's done and PR updated, pls review it at your convenience.

**Review by membphis — COMMENTED** [review]
missing test case to confirm the file `.githash` contains the correct content

**Comment by imjoey** [comment]
> missing test case to confirm the file `.githash` contains the correct content

@membphis thanks for your advice. PR updated and added the content verification in `.githash` file. Any feedback is welcome. Thank you.

**Comment by imjoey** [comment]
@juzhiyuan Hi, I notice that you moved the target branch to `v2.3` from `master` branch for this PR. Actually, this PR can only apply to `master` branch, since the content of  `./api/VERSION` in this PR is set to `master`. 

How about still keep the target branch as `master` for this PR. Once this PR is merged, I will proposed another PR to get along with `v2.3` branch for that new release? Looking forward to your response. Thanks.

**Comment by imjoey** [comment]
@membphis @nic-chen @juzhiyuan this pr is ready for reviewing, looking forward to your further suggestions. BTW, CI seems a little wired. Could I do anything to re-trigger it? Thanks.

**Comment by liuxiran** [comment]
It seems that only push can retigger ci, try to push -f?

**Comment by imjoey** [comment]
> It seems that only push can retigger ci, try to push -f?

@liuxiran thanks for reply. It seems that I've no additional changes to push. So let me wait for the `Update Branch` button. BTW, happy new year.

File: api/build.sh [hunk 2]
```diff
@@ -20,6 +20,7 @@ export ENV=local
 pwd=`pwd`
 
 VERSION=$(cat ./api/VERSION)
+GITHASH=$(cat ./.githash 2> /dev/null || git log --pretty=format:"%h" -1)
```
Comment (nic-chen) [inline api/build.sh:23]: There is still a risk of failure. It is recommended to remove the git dependencies.

Comment (imjoey) [reply api/build.sh]: @nic-chen good catch. Thanks. Especially in docker environment, the dependency of `git` command would definitely increase the size of image. Inspired by https://stackoverflow.com/questions/949314/how-to-retrieve-the-hash-for-the-current-commit-in-git , I borrow the awesome script here. Test passed in my local environment. Any feedback is welcome. Thanks.

Comment (imjoey) [reply api/build.sh]: @nic-chen  BTW, for `git` command is already widely used in `Makefile`, so I would prefer to keep using `git` to generate git hash into `.githash`. Please let me know if I was wrong. Thanks.

Comment (juzhiyuan) [reply api/build.sh]: cc @membphis Any ideas?

**Review by nic-chen — APPROVED** [review]
LGTM

**Review by juzhiyuan — APPROVED** [review]

**Pull request merged**